static/client/components/EditProductPanel/ProductActionModal.tsx

```diff
@@ -0,0 +1,127 @@
+import { useCallback, useState, type ReactNode } from "react";
+
+import { Modal, Button, Input, useNotify, Spinner } from "@canonical/react-components";
+import { useQueryClient } from "react-query";
+
+import { ProductsServices } from "@/services/api/services/products";
+import type { IProduct, IProductAction } from "@/services/api/types/products";
+
+interface ProductActionModalProps {
+  product: IProduct | null;
+  onClose: () => void;
+  action: IProductAction;
+  closeProductPanel: () => void;
+}
+
+const ProductActionModal = ({ product, onClose, action, closeProductPanel }: ProductActionModalProps): ReactNode => {
+  const [isLoading, setIsLoading] = useState(false);
+  const [newName, setNewName] = useState(product?.name || "");
+  const [inputError, setInputError] = useState<string | null>(null);
+  const queryClient = useQueryClient();
+
+  const notify = useNotify();
+  const handleAction = useCallback(async () => {
+    setIsLoading(true);
+    try {
+      if (action === "edit" && product) {
+        await ProductsServices.editProduct(product.id, newName);
+      } else if (action === "delete" && product) {
+        await ProductsServices.deleteProduct(product.id);
+      } else if (action === "add") {
+        await ProductsServices.addProduct(newName);
+      }
+
+      await queryClient.invalidateQueries("products");
+      await queryClient.refetchQueries("products");
```

Comment:
Do we need both? Usually when the query is invalidated, it will be refetched the next time it is rendered or immediately, if it is currently mounted.

```suggestion
      await queryClient.invalidateQueries("products");
```